### ISSUE TITLE
QUESTION: how to implement and call onPageRendered()?

### ISSUE THREAD

**Issue by randyparis** [stephanrauh/ngx-extended-pdf-viewer#2137]
Trying to speed up my pdf load by not waiting for all pages to get rendered before starting to display. onPageRendered() doesn't seem to be getting called - can't see where it's supposed to be getting called from...
Used example from PdfShowcase (https://pdfviewer.net/extended-pdf-viewer/prerendering) but it's not working: Put this code from showcase into component:
```ts
public spreadMode: 'off' | 'even' | 'odd' = 'off';
  public page: any;
// the above wasn't in the sample code, so I provided it..
// the below the sample code

  constructor(private pdfViewerService: NgxExtendedPdfViewerService) {
    pdfDefaultOptions.ignoreDestinationZoom = true;
  }

  public onPageRendered(): void {
    if (!this.pdfViewerService.isRenderQueueEmpty()) {
      // try again later when the pages requested by the pdf.js core or the user have been rendered
      setTimeout(() => this.onPageRendered(), 100);
    }

    const pagesBefore = this.spreadMode === 'off' ? 2 : 2;
    const pagesAfter = this.spreadMode === 'off' ? 2 : 5;
    let startPage = Math.max(this.page - pagesBefore, 1);
    let endPage = Math.min(this.page  + pagesAfter, this.pdfViewerService.numberOfPages());

    const renderedPages = this.pdfViewerService.currentlyRenderedPages();

    for (let page = startPage; page <= endPage; page++) {
      const pageIndex = page - 1;
      if (!this.pdfViewerService.hasPageBeenRendered(pageIndex)) {
          this.pdfViewerService.addPageToRenderQueue(pageIndex);
          console.log ('pageindex=' + pageIndex); // this is to see if it's being called and what it's reporting (mine)
          break; // break because you can request only one page at a time
      }
    }
  }
```
And this into <ngx-extended-pdf-viewer:
```html
(pageRendered)="onPageRendered()"
  [(page)]="page"
```
Not getting the console.log message, so I'm assuming it's not getting called

**Comment by stephanrauh** [comment]
As for rendering faster: you don't need to do anything. If range requests are configured correctly, the PDF viewer renders while still loading. If it works, the network tab of your browser's developer tools looks like so:
<img width="723" alt="image" src="https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/3045767/dc00de45-2c49-4639-aac9-f9316d0d4f22">

By default it loads chunks of 64 KB. The screenshot shows a larger chunk size because I've configured it in the demo.

**Comment by stephanrauh** [comment]
What I tried to show in the pre-rendering demo is something different: it shows how to render the next pages while the user is still reading a page. In most cases, the PDF viewer is fast enough so you don't need this, but some people have very complex files, so pre-rendering improves smooth scrolling.

**Comment by stephanrauh** [comment]
If your source code looks like the snippets, `onPageRendered()` should be called. Did you add a breakpoint or a `console.log()` at the first line of the method? BTW, a surprisingly elegant way to add a breakpoint is inserting the `debugger;` statement. I'm using it all the time, it's often easier than opening the file in the developer tools and finding the line of interest.

The demo https://pdfviewer.net/extended-pdf-viewer/pages-loaded shows that `(pageRendered)` is not broken. Maybe there's a small typo in your code.

**Comment by randyparis** [comment]
The below is what the network shows for my load: The actual pdf seems to be loading as your image shows, but until it does, the viewer takes a long tome to build and when it appears it displays a broken image link and text  “miniature of the page” until it resolves ( pdf.worker-4.0.731.min.js), and then the load progress zips to 100%.. lot of waiting until then

https://www.randyparis.com/RPArchive_update/assets/writing/manuscript_pdfs/my-new-pdfviewer/browser/?pdf=ARK2￼

￼

**Comment by stephanrauh** [comment]
Yes. That's puzzling me, too. The PDF viewer doesn't do much until the `pdf.worker` file has finished loading. For some reason, it doesn't even start on my machine. In any case, I'm very confused why it take 23 seconds to load the 740 KB size main.js.

<img width="846" alt="image" src="https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/3045767/db6ab4e4-88e4-445c-96f6-cc5615608ed5">

I wonder if we should do an experiment. If you send me the source code or the binaries of your application, I could temporarily deploy it on one of my dormant domains. This way, we could find out whether your domain suffers from a slow internet connection.

**Comment by randyparis** [comment]
Do yu want a zip of the dist build, or the un-processed original code?

**Comment by randyparis** [comment]
The zip of the dist file is 118.4 MB. a zip of the source code is 231.5 MB. neither can be emailed. Do you have a ftp server I could upload one or both too? Or I could put on mine for you to download.

Randy

**Comment by randyparis** [comment]
Ok. I think I’ve created a repository on git and invited you to collaborate. This is new for me.

it's under randyparis / my-new-pdfviewer-copy-ngx-loading-copy4

next time i’ll use a simpler name…

Please let me know if you can access it.

and THANKS

randy

**Comment by stephanrauh** [comment]
Yes, I've received your invitation and started to browse your repo. At the moment, it looks like the default project generated by `ng new my-new-pdfviewer-copy-ngx-loading-copy4`. I don't see any PDF file, nor do I see any usage of ngx-extended-pdf-viewer. Am I missing something?

**Comment by randyparis** [comment]
I’ve obviously done something wrong using the desktop git app. As I said, I’m new to actually trying to use git for anything other than reference.

On my desktop it let me open the project folder with all its assets, etc in the VS instance it provided, and I thought I sent it, but it looks like all git got was some kind of basic shell which doesn’t at all correlate.

is there a way to delete and start over?

**Comment by stephanrauh** [comment]
Well, now it worked!

**Comment by randyparis** [comment]
weird

**Comment by stephanrauh** [comment]
I'd like to close this issue. By now, we've established another channel to discuss your application, and I don't think our discussion adds much insight to the ngx-extended-pdf-viewer community. It's an interesting discussion, but that's another story. :)

Best regards,
Stephan
